src/commands/ScreenshotCommand.cpp

```diff
@@ -107,10 +107,7 @@ ScreenshotCommand::ScreenshotCommand()
    mbBringToTop=true;
    mIgnore=NULL;
    
-   static std::once_flag flag;
-   std::call_once( flag, []{
-      ::SetVetoDialogHook( MayCapture );
-   });
+   static VetoDialogHook::Scope scope{ MayCapture };
```

Comment:
Since `Scope` keeps the previous value it may hold some unnecessary data longer than one may expect, also the previous value will be restored when the static object is destroyed.

Reply:
I defined `Scope` to work generally with `GlobalVariable` specializations other than for functions.

In this case, the default-constructed empty `std::function` (which I expect is implemented as only a null pointer) is all that gets moved into the `scope` object, so it's only "wasting" one statically allocated pointer.

I allow `GlobalVariable` specialization to support non-scoped reassignments, but not by default.  I want the RAII idiom to apply be default.

I'm not sure if you are questioning whether the destructor of `scope` might come after the destruction of the local static variable in `Instance`, using that variable after destruction?  But I asked myself that question, and found a reassuring answer here:

https://en.cppreference.com/w/cpp/utility/program/exit

"if a function-local (block-scope) static object was initialized during construction of a subobject of a class or array, it is only destroyed after all subobjects of that class or all elements of that array were destroyed."

(And "subobject" is defined here https://en.cppreference.com/w/cpp/language/object to include member objects.)

That describes `instance` in GlobalVariable.h, which will be constructed (if not already constructed) when `scope` initializes its member `m_previous`.  Therefore `instance` will be destroyed, safely, only after `scope`.

Notice that I also provide `Initializer` for other cases where this guarantee does not apply, but this case doesn't need it for safety.